larq_compute_engine/core/bgemm_kernels_arm64.h

```diff
@@ -10,6 +13,94 @@ using compute_engine::core::TBitpacked;
+#define RUY_OFFSET_DST_COLS 44
+#define RUY_OFFSET_LHS_STRIDE 48
+#define RUY_OFFSET_RHS_STRIDE 52
+#define RUY_OFFSET_DST_STRIDE 56
+#define RUY_OFFSET_DEPTH 60
+#define RUY_OFFSET_OUTPUT_TRANSFORM_CLAMP_MIN 64
+#define RUY_OFFSET_OUTPUT_TRANSFORM_CLAMP_MAX 68
+#define RUY_OFFSET_OUTPUT_TRANSFORM_MULTIPLIER 72
+#define RUY_OFFSET_OUTPUT_TRANSFORM_BIAS 80
+#define RUY_OFFSET_OUTPUT_TRANSFORM_THRESHOLDS 64
+
+template <typename Params>
+void CheckOffsetsInBinaryKernelParams(const Params&) {
+  static_assert(offsetof(Params, lhs_base_ptr) == RUY_OFFSET_LHS_BASE_PTR, "");
+  static_assert(offsetof(Params, rhs_base_ptr) == RUY_OFFSET_RHS_BASE_PTR, "");
+  static_assert(offsetof(Params, dst_base_ptr) == RUY_OFFSET_DST_BASE_PTR, "");
+  static_assert(offsetof(Params, start_row) == RUY_OFFSET_START_ROW, "");
+  static_assert(offsetof(Params, start_col) == RUY_OFFSET_START_COL, "");
+  static_assert(offsetof(Params, last_row) == RUY_OFFSET_LAST_ROW, "");
+  static_assert(offsetof(Params, last_col) == RUY_OFFSET_LAST_COL, "");
+  static_assert(offsetof(Params, lhs_stride) == RUY_OFFSET_LHS_STRIDE, "");
+  static_assert(offsetof(Params, rhs_stride) == RUY_OFFSET_RHS_STRIDE, "");
+  static_assert(offsetof(Params, dst_stride) == RUY_OFFSET_DST_STRIDE, "");
+  static_assert(offsetof(Params, depth) == RUY_OFFSET_DEPTH, "");
+
+  const std::size_t OT_OFFSET = offsetof(Params, output_transform);
+
+  // For float output
+  static_assert(OT_OFFSET + offsetof(OutputTransform<float>, clamp_min) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_CLAMP_MIN,
+                "");
+  static_assert(OT_OFFSET + offsetof(OutputTransform<float>, clamp_max) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_CLAMP_MAX,
+                "");
+  static_assert(OT_OFFSET + offsetof(OutputTransform<float>, multiplier) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_MULTIPLIER,
+                "");
+  static_assert(OT_OFFSET + offsetof(OutputTransform<float>, bias) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_BIAS,
+                "");
+
+  // For int8 output
+  static_assert(OT_OFFSET + offsetof(OutputTransform<std::int8_t>, clamp_min) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_CLAMP_MIN,
+                "");
+  static_assert(OT_OFFSET + offsetof(OutputTransform<std::int8_t>, clamp_max) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_CLAMP_MAX,
+                "");
+  static_assert(
+      OT_OFFSET + offsetof(OutputTransform<std::int8_t>, multiplier) ==
+          RUY_OFFSET_OUTPUT_TRANSFORM_MULTIPLIER,
+      "");
+  static_assert(OT_OFFSET + offsetof(OutputTransform<std::int8_t>, bias) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_BIAS,
+                "");
+
+  // For bitpacked output
+  static_assert(OT_OFFSET + offsetof(OutputTransform<TBitpacked>, thresholds) ==
+                    RUY_OFFSET_OUTPUT_TRANSFORM_THRESHOLDS,
+                "");
+}
+
+#define MAKE_ZERO(reg) "eor " #reg ".16b, " #reg ".16b, " #reg ".16b\n"
```

Comment:
Is this faster than `"dup " #reg ".4s, wzr`?

Reply:
Yeah, because NEON isn't in general optimised for transferring from general registers into NEON registers.

From the [Cortex A72 software optimisation guide](https://developer.arm.com/documentation/uan0016/a/):

`dup` from a general register has a latency of 8 cycles and a throughput of 1 instruction per cycle.

![image](https://user-images.githubusercontent.com/7688302/92620720-dfc09300-f2ba-11ea-8065-c421282d0c24.png)

`eor` has a latency of 3 cycles and a throughput of 2 instructions per cycle.

![image](https://user-images.githubusercontent.com/7688302/92620775-f535bd00-f2ba-11ea-9d7e-088608280739.png)

Reply:
Nice find :+1: